docs/design/branch-mode.md

```diff
@@ -0,0 +1,117 @@
+# Branch mode
+
+This is a proposed configuration option to replace
+`experimental-advance-branches`. It is not intended to replace or obviate
+topics.
+
+The design below assumes that the configuration option is enabled. There should
+be no change from current behavior if the option is not enabled.
+
+**"Non-headless"** or **"branch"** mode: a git-colocated repository with the
+new configuration option enabled is considered to be in this mode whenever
+`.git/HEAD` (hereafter `HEAD`) is populated with a valid `jj` bookmark, and the
+bookmark is currently at `@` or `@-`. In this design, the bookmark will be
+referred to as `b(HEAD)`.
+
+**Note:** The only part of this design that is specific to `git` colocation is
+the use of `.git/HEAD`. The design could be implemented for any backend that
+has a concept of a "current" branch.
```

Comment:
I think that if we make this work, it should work on non-colocated repos as well. To me, this looks like a proposal of "what if we added the notion of a 'current branch' to `jj`?". TBH, I'm not sure I'm excited about this -- I'd probably want this to be optional and to keep it off by default -- but it is probably worth thinking through, and my first impression might be wrong.

This makes more sense to me than the experimental advance-branches feature in some ways, but it would also be a much larger change to implement. E.g., you'd need to adjust the log template to show which branch is current, store this information in the repo, etc.

Reply:
There's also something to be said for fixing the advance-branches feature, if it's possible without storing extra info (which I'm not at all sure about).

I think your original idea was to store the extra info in the Git repo only, but I'm suspicious about this part of it. We probably want it to work with `undo`, and so it must be stored on the `jj` side as well.